namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -32,80 +31,90 @@ class DnsSrvNamer(
+        lookup.setResolver(resolver)
+        Stat.time(latency)(lookup.run())
+        lookup.getResult match {
+          case DNS.Lookup.HOST_NOT_FOUND | DNS.Lookup.TYPE_NOT_FOUND =>
+            val msg = s"no results for $dsnsrvRecord"
+            log.trace(msg)
+            failure.incr()
+            update.update(Activity.Failed(new DNSLookupException(msg)))
+          case DNS.Lookup.SUCCESSFUL =>
+            val answers = Option(lookup.getAnswers).getOrElse(Array.empty)
+            val srvRecords = answers.flatMap {
+              case srv: DNS.SRVRecord => try {
+                val inetAddress = InetAddress.getByName(srv.getTarget.toString())
+                Some(Address(new InetSocketAddress(inetAddress, srv.getPort)))
+              } catch {
+                case _: UnknownHostException =>
+                  log.warning(s"srv lookup of $dsnsrvRecord returned unknown host ${srv.getTarget}")
+                  badHosts.incr()
+                  None
+              }
+              case _ => None
+            }
+            if (srvRecords.isEmpty) {
+              // valid DNS entry, but no instances.
+              // return NameTree.Neg because NameTree.Empty causes requests to fail,
+              // even in the presence of load-balancing (NameTree.Union) and fail-over (NameTree.Alt)
+              val msg = s"empty response for $dsnsrvRecord"
+              log.trace(msg)
+              zeroResults.incr()
+              update.update(Activity.Failed(new DNSLookupException(msg)))
+            } else {
+              log.trace("got %d results for %s", srvRecords.length, dsnsrvRecord)
+              success.incr()
+              update.update(Activity.Ok(srvRecords.toList))
             }
-            case _ => None
-          }
-          if (srvRecords.isEmpty) {
-            // valid DNS entry, but no instances.
-            // return NameTree.Neg because NameTree.Empty causes requests to fail,
-            // even in the presence of load-balancing (NameTree.Union) and fail-over (NameTree.Alt)
-            log.trace("empty response for %s", address)
-            zeroResults.incr()
-            NameTree.Neg
-          } else {
-            log.trace("got %d results for %s", srvRecords.length, address)
-            success.incr()
-            NameTree.Leaf(Name.Bound(Var.value(Addr.Bound(srvRecords: _*)), id, residual))
-          }
-        case code =>
-          val msg = s"unexpected result: $code for $address: ${lookup.getErrorString}"
-          log.error(msg)
-          throw new IOException(msg)
+          case code =>
+            val msg = s"unexpected result: $code for $dsnsrvRecord: ${lookup.getErrorString}"
+            log.error(msg)
+            update.update(Activity.Failed(new DNSLookupException(msg)))
+        }
+      }
+
+      doUnit()
+
+      timer.schedule(refreshInterval) {
+        doUnit()
       }
     }
+    Activity(state)
   }
 }
+
+case class DNSLookupException(msg: String)
```

Comment:
This looks good to me – I might consider having separate subtypes for the "lookup errored" and "lookup contained no SRV records" cases, but that's bikeshedding at this point... :)